pyroute2/ipset.py

```diff
@@ -196,7 +203,33 @@ def _entry_to_data_attrs(self, entry, etype, family):
             else:
                 raise TypeError('unknown family')
         for e, t in zip(entry.split(','), etype.split(',')):
-            if t in ('ip', 'net'):
+            if t == 'port':
+                if etype != 'port' and ':' in e:
+                    # Specifying protocol is only optional
+                    # in hash stypes
+                    proto, e = e.split(':')
+                else:
+                    proto = ''
+
+                if '-' in e:
+                    to, f = [int(x) for x in sorted(e.split('-'))]
+                    attrs += [
+                        ['IPSET_ATTR_PORT_FROM', f],
+                        ['IPSET_ATTR_PORT_TO', to]
+                    ]
+                else:
+                    attrs += [['IPSET_ATTR_PORT_FROM', int(e)]]
+
+                if etype != 'port':
+                    proto_map = {
+                        '': 1, # default value TCP
+                        'tcp': 1,
+                        'udp': 12,
+                    }
+                    attrs += [['IPSET_ATTR_PROTO', {'attrs': 
+                        [['IPSET_ATTR_NAME', 'x'*proto_map[proto]]]
```

Comment:
I am not happy with this approach, if someone could let me know how to better improve this, ie construct the array using the appropriate flags

Reply:
Could you please give an example of the target NLA structure? Looks a bit odd, that's true.

Reply:
It should look like:
{'attrs': [('IPSET_ATTR_IP_FROM', 
                    {'attrs': [('IPSET_ATTR_IPADDR_IPV4', '192.168.1.1')]}, 32768), 
                    ('IPSET_ATTR_PORT_FROM', 53, 16384), 
                    ('IPSET_ATTR_PROTO', {'attrs': []})]}, 32768)]}, 32768)],

I am unsure of how to express the protocol(UDP, TCP) as the arguments that need to be passed in.

Reply:
I should have also provided the struct from ipset:
`
struct hash_ipport4_elem {
	__be32 ip;
	__be16 port;
	u8 proto;
	u8 padding;
};
`

Reply:
I'll try to review the whole module on the weekend. Maybe it's time to refactor it a bit.

And maybe we should plan some IRC-meeting next week regarding this topic.

@ffourcot what do you think?

Reply:
Happy to have a chat, just shoot me through some details. In regards to a refactor, I have already written a wrapper around around the IPSet class, which I think would make a good base and happy to open another PR for it to be reviewed.

I was hoping to get this PR merged in as I need it from a project I am working on. But I am having issues with setting the protocol type of the ipset port added. Not sure how to express the the protocol as a NLA attribute. 

I manage to hack something together, to get it working. When the string passed in has a length of 1, it corresponds to TCP, when string length is 14 it corresponds to UDP. Not actually sure how to determine the correct type that I need to pass in order to do this properly. You able to point me in the right direction?

Reply:
Looks like there is a mistake in the NLA structure in the code you provided — I looked into the kernel, and `IPSET_ATTR_PROTO` should be `NLA_U8`, not a nested structure. So the kernel interpreted the nested NLA length as a value of `IPSET_ATTR_PROTO`.

Could you please provide me the bash code to do the same? With the ipset tool?

Reply:
@svinota @AZaugg : it's a work in progress, but what do you think of https://github.com/ffourcot/pyroute2/commits/portrange ? protocol support is planned but not done. That should be to complicated to add

Reply:
I was thinking of keeping with the core data types(int and str)  and was going to control them via new classes, example:
https://gist.github.com/AZaugg/c5c0648b8e87f09a5255de2e909b4503

`a = Hash('ipsetname, 'ip')
a.create()
a.add('192.168.0.1')`